UefiPayloadPkg/Library/PciSegmentInfoLibAcpiBoardInfo/PciSegmentInfoLibAcpiBoardInfo.c

```diff
@@ -55,18 +57,53 @@ RetrieveMultiSegmentInfoFromHob (
   mPciSegments = (PCI_SEGMENT_INFO *)AllocatePool (Size);
   ASSERT (mPciSegments != NULL);
   ZeroMem (mPciSegments, PciRootBridgeInfo->Count * sizeof (PCI_SEGMENT_INFO));
+  pPciSegments = (PCI_SEGMENT_INFO *)AllocatePool (Size);
+  ASSERT (pPciSegments != NULL);
+  ZeroMem (pPciSegments, PciRootBridgeInfo->Count * sizeof (PCI_SEGMENT_INFO));
 
   //
   // Create all root bridges with PciRootBridgeInfoHob
   //
   for (Index = 0; Index < PciRootBridgeInfo->Count; Index++) {
     if (UplSegmentInfo->SegmentInfo[Index].SegmentNumber == (UINT16)(PciRootBridgeInfo->RootBridge[Index].Segment)) {

```

Comment:
FdtParserLib will create both PciRootBridgeInfo HOB and UplSegmentInfo Hob following the same RbIndex, so it should be 1:1 mapping each other. Basing on this I think we do not need to compare SegmentNumber here, can you double check?